.github/workflows/continuous-integration-workflow.yml

```diff
@@ -7,12 +7,42 @@ on:
       - 'README.md'
 
 jobs:
-  build:
+  build-and-test:
     name: Build and test
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
     - name: build docker container 
       run: docker build --target test --tag todo-app:test .
     - name: run docker tests
-      run: docker run todo-app:test
\ No newline at end of file
+      run: docker run todo-app:test
+      
+  build-and-push-docker-image:
+    name: Docker Build, Tag & Publish
+    needs: build-and-test
+    if: ${{ github.ref == 'refs/heads/main' }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout Code
+        uses: actions/checkout@v1
+        id: git_checkout
+
+      - name: Build, Tag & Publish Image to DockerHub
+        uses: dinushchathurya/build-tag-push-action@v1.2.1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+          context: .
+          file: Dockerfile
+          repoOwner: ${{ secrets.DOCKERHUB_USERNAME }}
+          repository:  todo-app
+          tag: ${{github.ref_name}}
+
+  refresh-azure-image:
+    name: Refresh Azure Image
+    needs: build-and-push-docker-image
+    runs-on: ubuntu-latest
+    steps:
+      - name: Curl Azure webhook
+        run: |
+          curl --fail -dH -X POST '${{ secrets.AZURE_WEBHOOK }}'        
```

Comment:
I _believe_ this will work, but I can't see any evidence it's ever actually run - and it does sometimes report success even when it fails if there are mistakes in escaping the environment variable (given that you're using single-quotes, I would expect you to need to *not* escape the $ in the actual secret) - so worth checking this after it runs and making sure it outputs a sensible looking response (it should clearly include some output referencing your Azure app)